source/simulator/solver_schemes.cc

```diff
@@ -882,13 +912,23 @@ namespace aspect
     // Now iterate out the nonlinearities.
     dcr.stokes_residuals = std::pair<double,double>  (numbers::signaling_nan<double>(),
                                                       numbers::signaling_nan<double>());
-    for (nonlinear_iteration = 0; nonlinear_iteration < max_nonlinear_iterations; ++nonlinear_iteration)
+
+    SolverControl nonlinear_solver_control(max_nonlinear_iterations,
+                                           parameters.nonlinear_tolerance);
+
+    SolverControl nonlinear_solver_control_picard(newton_handler->parameters.max_pre_newton_nonlinear_iterations,
+                                                  newton_handler->parameters.nonlinear_switch_tolerance);
+
+    double relative_residual = std::numeric_limits<double>::max();
+    nonlinear_iteration = 0;
+
+    while (nonlinear_solver_control.check(nonlinear_iteration, relative_residual) == SolverControl::iterate)
```

Comment:
why is this one a while loop while you use do ... while in all other places?

Reply:
Because a) it was that way before as well (it was a for loop before), and b) we have some tests that assume this is a while loop (they set 0 max nonlinear iterations and expect to never go into the loop) and other tests that require all other schemes to not be a while loop (they require that the check is only executed after the nonlinear solve, and the number of nonlinear residuals recorded is 1 after 1 solve). I could make it work with consistent do .. while loops if I added a special condition that skipped the loop if max_nonlinear_iterations == 0, would you prefer that instead to keep the loops consistent?

Reply:
I would think setting the nonlinear iterations to 0 is a bug and should fail. What do you think?

Reply:
Is that the total number of nonlinear steps (Picard + Newton)? That seems wrong -- what is the solution vector supposed  to be if that happens?